### PR TITLE
Fix #6124 : Dropdown navigation like PrimeVue

### LINKED ISSUE #6124: Dropdown: keyboard navigation doesn't work on dropdown with groups
### Describe the bug

Keyboard navigation doesn't work on Dropdown with groups because `findNextOptionIndex` and `findPrevOptionIndex` uses top level of `visibleOptions` array to find next / prev position but in case of groups that array consists of group items and actual items are inside `items` property.

### Reproducer

https://primereact.org/dropdown/#group

### PrimeReact version

10.5.1

### React version

17.x

### Language

TypeScript

### Build / Runtime

Create React App (CRA)

### Browser(s)

_No response_

### Steps to reproduce the behavior

1. Open Dropdown with groups
2. Try to keyboard navigate with arrows

### Expected behavior

`findNextOptionIndex` and `findPrevOptionIndex` should look into `items` property for actual items.

### LINKED ISSUE #6417: Dropdown: All items at the same index in a group are marked focused
### Describe the bug

When using [grouped options](https://primereact.org/dropdown/#group), when hovering over an option, all options at the same index index in the group have the `p-focus` class and the `data-p-focused="true"` attribute set. See screenshot from the PrimeReact site:
Screenshot 2024-04-18 at 10 06 07 PM

### Reproducer

_No response_

### PrimeReact version

10.6.3

### React version

18.x

### Language

TypeScript

### Build / Runtime

Vite

### Browser(s)

_No response_

### Steps to reproduce the behavior

_No response_

### Expected behavior

_No response_

### PR DESCRIPTION
**Pull request by KirilCycle**
Fix #6124
Fix #6417 
Each Group item now have unique index

### CODE REVIEW

File: components/lib/dropdown/Dropdown.js [hunk 1]
```diff
@@ -352,22 +355,26 @@ export const Dropdown = React.memo(
             return selectedIndex < 0 ? findLastOptionIndex() : selectedIndex;
         };
 
+
+        const appropriateOptions = () => flatedGroupOptions ?? visibleOptions
```
Comment (KirilCycle) [inline components/lib/dropdown/Dropdown.js:359]: maybe its better to use a simple variable

**Comment by KirilCycle** [comment]
@melloware Hello! Its done

File: components/lib/dropdown/Dropdown.js (hunk 1 above)
Comment (melloware) [reply components/lib/dropdown/Dropdown.js:359]: Yes move this up to where you get `flattenedOptions`  and shouldn't `visibleOptions = flattenedOptions` and then all this code will just work using `visibleOptions`?

File: components/lib/dropdown/Dropdown.js [hunk 2]
```diff
@@ -57,6 +57,8 @@ export const Dropdown = React.memo(
             when: overlayVisibleState
         });
 
+        const flatedGroupOptions = props.optionGroupLabel ? props.options.flatMap((item) => item.items) : null
```
Comment (melloware) [inline components/lib/dropdown/Dropdown.js:60]: This should be `flattenedOptions`

File: components/lib/dropdown/DropdownPanel.js [hunk 3]
```diff
@@ -76,7 +76,7 @@ export const DropdownPanel = React.memo(
             }
         };
 
-        const createGroupChildren = (optionGroup, style) => {
+        const createGroupChildren = (optionGroup, style, passedOptionsCount) => {
```
Comment (melloware) [inline components/lib/dropdown/DropdownPanel.js:79]:
```suggestion
        const createGroupChildren = (optionGroup, style, index) => {
```

Call `passedOptionsCount` just `index`

**Review by melloware — CHANGES_REQUESTED** [review]

File: components/lib/dropdown/Dropdown.js (hunk 1 above)
Comment (KirilCycle) [reply components/lib/dropdown/Dropdown.js:359]: visibleOptions need to be the same as it was, this format will be used at DropdownPanel component to make groups

**Comment by melloware** [comment]
@KirilCycle i think this may also fix this issue correct?  https://github.com/primefaces/primereact/issues/6417

**Comment by KirilCycle** [comment]
> @KirilCycle i think this may also fix this issue correct? #6417

Exactly!

File: components/lib/dropdown/Dropdown.js (hunk 1 above)
Comment (KirilCycle) [reply components/lib/dropdown/Dropdown.js:359]: I cant  access visibleOptions  before initialization, should i leave it as method or move up their initialization ?

**Comment by melloware** [comment]
Let me take a deeper look at this. I like where you are going though.

**Comment by KirilCycle** [comment]
> Let me take a deeper look at this. I like where you are going though.

I also second approach, we can use simple number for key at simple list, but for group list we can use key like `${groupIndex}_${index}`

**Comment by melloware** [comment]
I think we need to study PrimeVue they actually do this: https://github.com/primefaces/primevue/blob/344ec64cd461aad1493073dd3ff10bc8a7338654/components/lib/dropdown/Dropdown.vue#L879-L888

And the Groups are actually a different type of Option that has `result.push({ optionGroup: option, group: true, index });` so they index everything properly in getVisibleOptions

https://github.com/primefaces/primevue/blob/344ec64cd461aad1493073dd3ff10bc8a7338654/components/lib/dropdown/Dropdown.vue#L902-L926

**Comment by KirilCycle** [comment]
> I think we need to study PrimeVue they actually do this: https://github.com/primefaces/primevue/blob/344ec64cd461aad1493073dd3ff10bc8a7338654/components/lib/dropdown/Dropdown.vue#L879-L888
> 
> And the Groups are actually a different type of Option that has `result.push({ optionGroup: option, group: true, index });` so they index everything properly in getVisibleOptions
> 
> https://github.com/primefaces/primevue/blob/344ec64cd461aad1493073dd3ff10bc8a7338654/components/lib/dropdown/Dropdown.vue#L902-L926

Should I try to use it the way it's done in vue?

**Comment by melloware** [comment]
Yes we should see if we can modify things the Vue way.   If you try the Vue showcase the Dropdown with Groups is working fine: https://primevue.org/dropdown/#group

**Comment by KirilCycle** [comment]
> Yes we should see if we can modify things the Vue way. If you try the Vue showcase the Dropdown with Groups is working fine: https://primevue.org/dropdown/#group

Okey ill take a look

**Comment by KirilCycle** [comment]
@melloware Vue Dropdown will flat all this to single options list first, and then all logic will work with this flatted list,And React Dropdown component use same format as it was passed and functions are not able to work with this type of data, and DropdownPanel is designed to handle groups the way they are, however still has an issue with indexses. Should i keep current solution to this problem or apply Vue behaviour ?

**Comment by melloware** [comment]
Yes I saw that. I feel like Vue is doing it a cleaner way and we just need to tweak the React code to do the same thing. I think DropDownPanel will need some changes but I think in the long run this will be better.

File: components/lib/dropdown/Dropdown.js [hunk 4]
```diff
@@ -57,15 +57,30 @@ export const Dropdown = React.memo(
...
+
             if (hasFilter && !isLazy) {
                 const filterValue = filterState.trim().toLocaleLowerCase(props.filterLocale);
                 const searchFields = props.filterBy ? props.filterBy.split(',') : [props.optionLabel || 'label'];
 
                 if (props.optionGroupLabel) {
                     let filteredGroups = [];
 
-                    for (let optgroup of props.options) {
+                    for (let optgroup of options) {
```
Comment (KirilCycle) [inline components/lib/dropdown/Dropdown.js:83]: I did not check if filter functionality works fine yet

File: components/lib/dropdown/DropdownPanel.js [hunk 5]
```diff
@@ -144,17 +113,14 @@ export const DropdownPanel = React.memo(
...
-                        {groupChildrenContent}
-                    </React.Fragment>
-                );
+                    </li>
+                )
+                
             }
-
-            const optionLabel = props.getOptionLabel(option);
             const optionKey = index + '_' + props.getOptionRenderKey(option);
```
Comment (KirilCycle) [inline components/lib/dropdown/DropdownPanel.js:122]: Group Item key not the same as it was, i do not use here group index to create render key, is this okey ?

File: components/lib/dropdown/Dropdown.js [hunk 6]
```diff
@@ -57,6 +57,9 @@ export const Dropdown = React.memo(
             when: overlayVisibleState
         });
 
+        const flattenedOptions = props.optionGroupLabel ? props.options.flatMap((item) => item.items) : null
```
[the remark this answers is not in the document]
Comment (KirilCycle) [reply components/lib/dropdown/Dropdown.js]: Sorry, what you mean ? Filtering works fine with last commit

File: components/lib/dropdown/DropdownPanel.js (hunk 5 above)
Comment (KirilCycle) [reply components/lib/dropdown/DropdownPanel.js:122]: Same thing, but the index is at the back

**Comment by melloware** [comment]
Wow this is looking so much cleaner!

**Comment by melloware** [comment]
Oh run `npm run format` on the code.

**Review by melloware — APPROVED** [review]

**Review by nitrogenous — APPROVED** [review]
LGTM. Amazing job thx!

**Pull request merged**